src/freertos_drivers/common/FileSystem.cxx

```diff
@@ -0,0 +1,366 @@
+ */
+int FileSystem::open(struct _reent *reent, const char *path, int flags, int mode)
+{
+    mutex.lock();
+    int fd = fd_alloc();
+    mutex.unlock();
+    if (fd < 0)
+    {
+        errno = EMFILE;
+        return -1;
+    }
+    files[fd].flags = flags;
+
+    FileSystem *fs = fs_lookup(path);
+    if (fs)
+    {
+        files[fd].dev = fs;
+        files[fd].device = false;
+        const char *subpath = path + strlen(fs->name) + 1;
+        int result = files[fd].dev->open(&files[fd], subpath, flags, mode);
+        if (result < 0)
+        {
+            fd_free(fd);
+            errno = -result;
+            return -1;
+        }
+        return fd;
+    }
+    // No device found.
+    fd_free(fd);
+    errno = ENODEV;
+    return -1;
+}
+
+/** Close a file or device.
+ * @param reent thread save reentrant structure
+ * @param fd file descriptor to close
+ * @return 0 upon success, -1 upon failure with errno containing the cause
+ */
+int FileSystem::close(struct _reent *reent, int fd)
+{
+    File* f = file_lookup(fd);
+    if (!f) 
+    {
+        /* errno should already be set appropriately */
+        return -1;
+    }
+    if (fd >=0 && fd <= 2)
+    {
+        // stdin, stdout, and stderr never get closed
+        return 0;
+    }
+    int result = f->dev->close(f);
+    if (result < 0)
+    {
+        errno = -result;
+        return -1;
+    }
+    fd_free(fd);
+    return 0;
+}
+
+/** Remove a file.
+ * @param reent thread safe reentrant structure
+ * @param path file name
+ * @return 0 upon success, -1 upon failure with errno containing the cause
+ */
+int FileSystem::unlink(struct _reent *reent, const char *path)
+{
+    FileSystem *fs = fs_lookup(path);
+    if (fs)
+    {
+        const char *subpath = path + strlen(fs->name) + 1;
+        int result = fs->unlink(subpath);
+        if (result < 0)
+        {
+            errno = -result;
+            return -1;
+        }
+        return 0;
+    }
+
+    errno = ENOENT;
+    return -1;
+}
+
+/** Get the status information of a file or device.
+ * @param reent thread save reentrant structure
+ * @param path file or device name
+ * @param stat structure to fill status info into
+ * @return 0 upon success, -1 upon failure with errno containing the cause
+ */
+int FileSystem::stat(struct _reent *reent, const char *path, struct stat *stat)
+{
+    FileSystem *fs = fs_lookup(path);
+    if (fs)
+    {
+        const char *subpath = path + strlen(fs->name) + 1;
+        int result = fs->stat(subpath, stat);
+        if (result < 0)
+        {
+            errno = -result;
+            return -1;
+        }
+        return 0;
+    }
+
+    errno = ENOENT;
+    return -1;
+}
+
+/** Get the status information of a file or device.
+ * @param file file reference for this device
+ * @param stat structure to fill status info into
+ * @return 0 upon successor or negative error number upon error.
+ */
+int FileSystem::fstat(File* file, struct stat *stat)
```

Comment:
can we delete this function?

Reply:
No, this is the default implementation.  Eventually this will be important in the refactor.